feat: allow import peppered/salted password hashes #2946

ci42 opened the pull request:
<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.

This text will be included in the changelog. If applicable, include links to documentation or pieces of code.
If your change includes breaking changes please add a code block documenting the breaking change:

```
Allow peppered/salted password hashes
```
-->

## Related issue(s)

<!--
If this pull request

1. is a fix for a known bug, link the issue where the bug was reported in the format of `#1234`;
2. is a fix for a previously unknown bug, explain the bug and how to reproduce it in this pull request;
3. implements a new feature, link the issue containing the design document in the format of `#1234`;
4. improves the documentation, no issue reference is required.

Pull requests introducing new features, which do not have a design document linked are more likely to be rejected and take on average 2-8 weeks longer to
get merged.

You can discuss changes with maintainers either in the Github Discussions in this repository or
join the [Ory Chat](https://www.ory.sh/chat).
-->
#2946 

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR.

Please be aware that pull requests must have all boxes ticked in order to be merged.

If you're unsure about any of them, don't hesitate to ask. We're here to help!
-->

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ory/kratos?pullRequest=3057) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ory/kratos?pullRequest=3057) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ory/kratos?pullRequest=3057) it.</sub>

zepatrik commented on embedx/config.schema.json:
```diff
@@ -2327,7 +2327,29 @@
       "additionalProperties": false
     },
-    "ciphers": {
+    "passwordhash_comparators": {
```
Why is this a separate top level config when we already have the `hashers` config? Can you add it there instead?

ci42 replied:
moved to the `hashers` config

zepatrik commented on hash/hash_comparator.go:
```diff
@@ -458,9 +624,6 @@ func decodeSSHAHash(encodedHash string) (hasher string, salt, hash []byte, err e
 		hasher = "sha512"
 		index_of_hash_begin = 9
 		index_of_salt_begin = 64
-
-	default:
-		return "", nil, nil, ErrInvalidHash
```
I don't think this should be removed, please revert.

zepatrik commented on hash/hasher_test.go:
```diff
@@ -251,19 +253,44 @@ func TestCompare(t *testing.T) {
 	assert.Error(t, hash.Compare(context.Background(), []byte("tesu"), []byte("$scrypt$ln=16384,r=8,p=1$2npRo7P03Mt8keSoMbyD/tKFWyUzjiQf2svUaNDSrhA=$MiCzNcIplSMqSBrm4HckjYqYhaVPPjTARTzwB1cVNYE=")))
 	assert.Error(t, hash.Compare(context.Background(), []byte("tesu"), []byte("$scrypt$ln=abc,r=8,p=1$2npRo7P03Mt8keSoMbyD/tKFWyUzjiQf2svUaNDSrhA=$MiCzNcIplSMqSBrm4HckjYqYhaVPPjTARTzwB1cVNYE=")))
 
+	assert.Nil(t, hash.Compare(context.Background(), []byte("test"), []byte("$md5$CY9rzUYh03PK3k6DJie09g==")))
```
Moving the block of md5 assertions makes it very hard to review, please move it back so the diff properly shows what changed.

zepatrik commented on hash/hash_comparator.go:
```diff
@@ -28,45 +28,106 @@ import (
-		return CompareMD5(ctx, password, hash)
-	default:
-		return errors.WithStack(ErrUnknownHashAlgorithm)
+var hashComparators []PasswordHashComparator
```
I don't think we need to refactor the package like this because there is already an established way to pass config to the hash comparators. Regardless of whether the code is nicer or not, in the end it makes it impossible to review with refactoring and new features mixed. Please use the existing way to pass the hash config, and revert all the refactoring. I am open to discuss that, but it should really be a separate PR.

ci42 replied:
> there is already an established way to pass config to the hash comparators.

Maybe I missed something, but other than the PHC string, I don't see any way to pass conf data in the existing approach. IMHO the PHC string is not the right place for pepper values (see the discussion [here](https://github.com/ory/kratos/issues/2946#issuecomment-1453139768)). So refactoring is a precondition and not just to make the code nicer.

zepatrik reviewed (changes requested):
Thanks for putting in the effort. however we need to fix a few things. It is also not a good idea to mix new features with refactoring in one PR, so I would suggest you either split this in two PRs, or we don't do the refactoring at all.
For pepper secrets, I think we need to include the possibility to rotate the secret as well. It is not cheap to do, as we need to hash the password multiple times, but there is no other way to update peppers or increase the secret entropy without a rotation mechanism, so this is a requirement IMO.

aeneasr commented:
See also feedback from https://github.com/ory/kratos/issues/2946#issuecomment-1451020646

ci42 commented on embedx/config.schema.json:
```diff
@@ -2327,7 +2327,29 @@
+        "required": ["Id"],
+        "type": "object",
+        "properties": {
+          "Id": {
```
done

ci42 commented:
@aeneasr @zepatrik
I have completely rewritten the PR. The new hash comparison function is now similar to the existing hash comparison functions. t now supports various old hashing methods (md5, sha1, sha256, sha512) as well as optional salts, peppers and a work factor (number of iterations). The old hash itself is hashed with argon2id for improved security (e.g. offline attacks in case of a database leak).

All parameters are encoded in the hash string. The hash string has the form: 
`$legacy$<legacy hash parameters>$<argon2id in phc format>` - where the legacy hash parameters are encoded as follows  
`alg=<md5|sha1|sha256|sha512>,format=<base64 encoded format, e.g: '{PEPPER}-{PWD}-{SALT}'>,i=<number of iterations>,(salt=<base64 encoded salt>,pepper=<base64 encoded pepper>)`

An example using a salted and peppered md5 in the format `{SALT}::{PWD}::{PEPPER}`with 3 iterations: `$legacy$alg=md5,format=e1NBTFR9Ojp7UFdEfTo6e1BFUFBFUn0,i=3,salt=7aCA,pepper=/v7//w$argon2id$v=19$m=65536,t=2,p=2$YuYcJII0fm8UXm8XiCqimA$xzUihMR5G40U3Gf2CmTfNhXDYTtva6ykbjGnfQnWCac`

zepatrik commented on hash/hash_comparator.go:
```diff
@@ -558,3 +576,78 @@ func decodeMD5Hash(encodedHash string) (pf, salt, hash []byte, err error) {
+	phc = "$" + parts[3]
+	legacyParams := strings.Split(parts[2], ",")
+	for _, lp := range legacyParams {
+		switch kv := strings.Split(lp, "="); kv[0] {
+		case "alg":
+			alg = kv[1]
```
Here is a potential for index out of bounds panics, I'd propose to use this instead:

```suggestion
		k, v, found := strings.Cut(lp, "=")
		if !found {
			return "", "", "", 0, nil, nil, ErrInvalidHash
		}
		switch k {
		case "alg":
			alg = v
```

zepatrik commented on hash/hash_comparator.go:
```diff
@@ -558,3 +576,78 @@ func decodeMD5Hash(encodedHash string) (pf, salt, hash []byte, err error) {
+
+// decodeLegacyHash decodes argon2id wrapped legacy password hash.
+// format: $legacy$alg=<md5|sha1|sha256|sha512>,format=<base64 encoded format, e.g: '{PEPPER}-{PWD}-{SALT}'>,i=<number of iterations>,(salt=<base64 encoded salt>,pepper=<base64 encoded pepper>)$<argon2id encoded hash>
+func decodeLegacyHash(encodedHash string) (phc, alg, format string, iterations int, salt, pepper []byte, err error) {
```
As is, this does not require any of the params to be set. Do we need to set some defaults, e.g. for the format? Or should we return an `ErrInvalidHash` in case some of the required params are missing?

zepatrik commented on hash/hash_comparator.go:
```diff
@@ -558,3 +576,78 @@ func decodeMD5Hash(encodedHash string) (pf, salt, hash []byte, err error) {
+}
+
+func legacyHash(alg, format string, iterations int, salt, pepper, pwd []byte) ([]byte, error) {
+	if len(pwd) < 1 || !strings.Contains(format, "{PWD}") || iterations < 1 || !(alg == "md5" || alg == "sha1" || alg == "sha256" || alg == "sha512") {
+		return nil, ErrInvalidHash
+	}
```
I guess validation is covered in this part.

ci42 replied:
yep

zepatrik commented on hash/hash_comparator.go:
```diff
@@ -558,3 +576,78 @@ func decodeMD5Hash(encodedHash string) (pf, salt, hash []byte, err error) {
+	if pepper != nil {
+		replacements = append(replacements, "{PEPPER}", string(pepper))
+	}
+	toHash := []byte(strings.NewReplacer(replacements...).Replace(format))
+
+	var hash []byte
+	for i := 1; i <= iterations; i++ {
+		switch alg {
+		case "md5":
+			sum := md5.Sum(toHash) //#nosec G401 -- compatibility for imported passwords
+			hash = sum[:]
+		case "sha1":
+			sum := sha1.Sum(toHash) //#nosec G401 -- compatibility for imported passwords
+			hash = sum[:]
+		case "sha256":
+			sum := sha256.Sum256(toHash)
+			hash = sum[:]
+		case "sha512":
+			sum := sha512.Sum512(toHash)
+			hash = sum[:]
+		default:
+			return nil, ErrUnknownHashAlgorithm
+		}
+		toHash = hash
+	}
+	return hash, nil
```
I think this is a bit easier to understand:

```suggestion
	toHash := []byte(strings.NewReplacer(replacements...).Replace(format))

	var hasher hash.Hash
	switch alg {
	case "md5":
		hasher = md5.New() //#nosec G401 -- compatibility for imported passwords
	case "sha1":
		hasher = sha1.New() //#nosec G401 -- compatibility for imported passwords
	case "sha256":
		hasher = sha256.New()
	case "sha512":
		hasher = sha512.New()
	default:
		return nil, ErrUnknownHashAlgorithm
	}
	for i := 1; i <= iterations; i++ {
		toHash = hasher.Sum(toHash)
	}
	return toHash, nil
```

ci42 replied:
`toHash = hasher.Sum(toHash)` does not do what you and I think 😄  Should be:
```
hasher.Write(toHash)
toHash = hasher.Sum([]byte{})
hasher.Reset()
```

zepatrik reviewed (commented):
Looks good now, thanks for revisiting!
I just have some minor suggestions, nothing critical.

ci42 commented:
@zepatrik Thanks for the review. I've included your suggested changes, added some more tests and rebased the PR onto the current master.

By the way, I made a [PR](https://github.com/ory/kratos/pull/3186) with some minor hash_comparator.go refactorings. Maybe you can have a look at it as well.

zepatrik reviewed (approved):
Nice :+1:
